[WIP] feat: add language selector

fsdiogo opened the pull request:
What's missing:

- [ ] Get all the available locales
- [ ] Show full language name instead of the short code


Closes https://github.com/ipfs-shipyard/ipfs-webui/issues/795.

fsdiogo commented:
![lang](https://user-images.githubusercontent.com/33324750/47228383-f6a32700-d3bc-11e8-9b67-c89984428934.gif)

Don't know why I only get `en` when querying the i18n languages.

fsdiogo commented:
Current status:

![i18n](https://user-images.githubusercontent.com/33324750/47659587-302e1c00-db8d-11e8-9dc1-d38e2c62ab6b.gif)

Why use modal instead of dropdown: https://github.com/ipfs-shipyard/ipfs-webui/issues/795#issuecomment-430580143

olizilla commented on src/settings/SettingsPage.js:
```diff
@@ -2,23 +2,43 @@ import React from 'react'
+    <Box className='mb3 pa4'>
+      <div className='flex justify-between items-center'>
+        <p className='ma0 lh-copy charcoal-muted f6'>
+          {t('language', { language: isoLangs[i18n.language].nativeName })}
```
`isoLangs[i18n.language].nativeName` is risky. It assumes that whatever comes back from i18n.language will definitly be a key the isoLangs object. 

As an example, `i18n.language` reports `en-GB` on my machine becuase I am so very british i have a locale for it. As such the settings page blows up for me.

fsdiogo replied:
😮 gotta figure out a work around for that then.

olizilla commented:
Why is `svenska` listed all lowercase in that gif? Is it all officially an all lowercase language?

fsdiogo commented:
We now have a `languages.json` that includes the language codes with the country variants.

Unfortunately the process is not automatic, when we add a new language we have to manually import it in the `i18n.js` and if it has a country variant we need to add it in the same file.

olizilla commented on src/settings/SettingsPage.js:
```diff
@@ -2,23 +2,40 @@ import React from 'react'
+    <Box className='mb3 pa4'>
+      <div className='flex justify-between items-center'>
+        <p className='ma0 lh-copy charcoal-muted f6'>
+          {t('language', { language: getCurrentLanguage() })}
```
UI suggestion, make the label look like our block labels, make the language name larger and more obvious... if you need to change it you're gonna need it big and clear, and bring the button next to it.

<img width="882" alt="screenshot 2018-10-31 at 11 06 03" src="https://user-images.githubusercontent.com/58871/47784330-23303a80-dcfd-11e8-9a5d-4371b276f332.png">

It could also be it's own compoent, that opens and closes the model, so the parent page doesn't need to know abou it.

olizilla commented on src/settings/language-modal/LanguageModal.js:
```diff
@@ -0,0 +1,52 @@
+        <p className='gray w-80 center'>{t('languageModal.description')}</p>
+        <div className='pa2 flex flex-wrap'>
+          { localesList.map((lang) =>
+            <button
```
For some reason the button styles need to include a background color to reset them on FF, and we need to give the text more space. Perhaps the modal could be wider.

<img width="872" alt="screenshot 2018-10-31 at 11 12 35" src="https://user-images.githubusercontent.com/58871/47784566-e4e74b00-dcfd-11e8-82f9-4930f9edb3b0.png">

fsdiogo commented:
Lastest commit introduces two changes:

- Updated the language selector UI
- Updated the IPFS Config layout from display table to flex and upgraded its responsiveness

## Language selector

![lang](https://user-images.githubusercontent.com/33324750/47795572-af9d2600-dd1a-11e8-84e2-b73f011b2477.gif)

## IPFS Config responsiveness

![resp](https://user-images.githubusercontent.com/33324750/47795774-0145b080-dd1b-11e8-8a63-bc2f815fcaf9.gif)

olizilla commented on src/lib/languages.json:
```diff
@@ -0,0 +1,433 @@
+{
+  "af-NA": "Afrikaans (Namibia)",
```
I think it'd be more friendly to display the languages in their local name, rather than english for all the things.

<img width="639" alt="screenshot 2018-11-02 at 12 09 34" src="https://user-images.githubusercontent.com/58871/47915046-8bb41e80-de99-11e8-958b-f388f1a9044d.png">

olizilla replied:
@lidel have you got any sources for languages in their local language. How do we deal with the multiple Chinese varients? I'm stuggling to find a simple process for this. I have pinged transifex.

olizilla replied:
More here https://github.com/ipfs/i18n/issues/4

lidel replied:
I was about to link https://github.com/mozilla/language-mapping-list/blob/master/language-mapping-list.js but you already found it :) (`nativeName` is what we want to show to end user)

olizilla commented:
woo haa

<img width="818" alt="screenshot 2018-11-02 at 17 48 46" src="https://user-images.githubusercontent.com/58871/47931738-a1d8d380-dec7-11e8-8cc3-e607ba97d178.png">

olizilla commented:
@lidel pls can you eyeball this when you get a moment

fsdiogo commented:
Can this be merged @olizilla @lidel?

The language code was being displayed when it wasn't available in the `languages.json`. In the latest commit if the language and country code isn't available, I fallback to showing only the language code if possible, if not I fallback to English because is the default language of the Web UI.

lidel commented on public/locales/en/settings.json:
```diff
@@ -3,6 +3,16 @@
   "save": "Save",
   "saving": "Saving...",
   "reset": "Reset",
+  "language": "Language",
```
Would it be helpful to clarify what is the target of language picker? 
For example:
```suggestion
  "language": "Web UI Language",
```

fsdiogo replied:
I don't think it's necessary. It's expected to change the Web UI language, as we are in the Web UI settings no? 🤔

olizilla replied:
I think `Language` is ok.

olizilla commented on src/components/language-selector/language-modal/LanguageModal.js:
```diff
@@ -0,0 +1,52 @@
+
+const LanguageModal = ({ t, tReady, onLeave, link, className, ...props }) => {
+  const handleClick = (lang) => {
+    console.log(lang)
```
rm `console.log(lang)`

olizilla commented on src/i18n.js:
```diff
@@ -22,18 +22,32 @@ import sv from 'i18next-icu/locale-data/sv'
     ns: ['welcome', 'status', 'files', 'explore', 'peers', 'settings', 'notify'],
-    fallbackLng: 'en',
+    fallbackLng: {
+      'zh-HANS': ['zh-CN', 'en'],
```
After much digging, i think `Hans` and `Hant` should only have the first letter capitalized

```suggestion
      'zh-Hans': ['zh-CN', 'en'],
```

olizilla replied:
See:
- Mozilla langmap https://github.com/mozilla/language-mapping-list/blob/6a1ba99f9625486705f0872750df228f52498bea/language-mapping-list.js#L837
- Spec http://www.ietf.org/rfc/bcp/bcp47.txt

olizilla commented on src/i18n.js:
```diff
@@ -22,18 +22,32 @@ import sv from 'i18next-icu/locale-data/sv'
-    fallbackLng: 'en',
+    fallbackLng: {
+      'zh-HANS': ['zh-CN', 'en'],
+      'zh-HANT': ['zh-TW', 'en'],
```
```suggestion
      'zh-Hant': ['zh-TW', 'en'],
```
